File: app/src/main/java/com/minerdev/greformanager/ImageListAdapter.java
package com.minerdev.greformanager;

import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Collections;

public class ImageListAdapter extends RecyclerView.Adapter<ImageListAdapter.ViewHolder> {
     ArrayList<Uri> items = new ArrayList<>();
     OnItemClickListener listener = new OnItemClickListener();
     int thumbnailPos = 0;

    public void setOnItemClickListener(OnItemClickListener clickListener) {
        listener = clickListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View itemView = inflater.inflate(R.layout.item_image, parent, false);
        return new ViewHolder(itemView, listener);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.setItem(items.get(position), thumbnailPos);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void addItem(Uri item) {
        items.add(item);
    }

    public Uri getItem(int position) {
        return items.get(position);
    }

    public void setItem(int position, Uri item) {
        items.set(position, item);
    }

    public void removeItem(int position) {
        items.remove(position);
    }

    public ArrayList<Uri> getItems() {
        return items;
    }

    public void setItems(ArrayList<Uri> items) {
        this.items = items;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
         final ImageView imageView;
         final TextView textView;

        public ViewHolder(@NonNull View itemView, final OnItemClickListener clickListener) {
            super(itemView);

            imageView = itemView.findViewById(R.id.imageItem_imageView);
            textView = itemView.findViewById(R.id.imageItem_textView_thumbnail);

            ImageButton imageButton_delete = itemView.findViewById(R.id.imageItem_imageButton_delete);
            imageButton_delete.setOnClickListener(v -> {
                if (clickListener != null) {
                    clickListener.onDeleteButtonClick(ViewHolder.this, itemView, getAdapterPosition());
                }
            });

            ImageButton imageButton_thumbnail = itemView.findViewById(R.id.imageItem_imageButton_thumbnail);
            imageButton_thumbnail.setOnClickListener(v -> {
                if (clickListener != null) {
                    clickListener.onThumbnailButtonClick(ViewHolder.this, itemView, getAdapterPosition());
                }
            });

            ImageButton imageButton_up = itemView.findViewById(R.id.imageItem_imageButton_up);
            imageButton_up.setOnClickListener(v -> {
                if (clickListener != null) {
                    clickListener.onUpButtonClick(ViewHolder.this, itemView, getAdapterPosition());
                }
            });

            ImageButton imageButton_down = itemView.findViewById(R.id.imageItem_imageButton_down);
            imageButton_down.setOnClickListener(v -> {
                if (clickListener != null) {
                    clickListener.onDownButtonClick(ViewHolder.this, itemView, getAdapterPosition());
                }
            });
        }

        public void setItem(Uri uri, int thumbnailPos) {
            imageView.setImageURI(uri);

            if (getAdapterPosition() == thumbnailPos) {
                textView.setVisibility(View.VISIBLE);

            } else {
                textView.setVisibility(View.GONE);
            }
        }
    }

    public class OnItemClickListener {
        public void onDeleteButtonClick(ImageListAdapter.ViewHolder viewHolder, View view, int position) {
            if (thumbnailPos == position) {
                thumbnailPos = 0;

            } else if (thumbnailPos > position) {
                thumbnailPos--;
            }

            removeItem(position);
            notifyDataSetChanged();
        }

        public void onThumbnailButtonClick(ImageListAdapter.ViewHolder viewHolder, View view, int position) {
            thumbnailPos = position;
            notifyDataSetChanged();
        }

        public void onUpButtonClick(ImageListAdapter.ViewHolder viewHolder, View view, int position) {
            if (position > 0) {
                Collections.swap(items, position - 1, position);
                notifyDataSetChanged();

                if (thumbnailPos == position) {
                    thumbnailPos--;

                } else if (thumbnailPos == position - 1) {
                    thumbnailPos++;
                }
            }
        }

        public void onDownButtonClick(ImageListAdapter.ViewHolder viewHolder, View view, int position) {
            if (position < items.size() - 1) {
                Collections.swap(items, position, position + 1);
                notifyDataSetChanged();

                if (thumbnailPos == position) {
                    thumbnailPos++;

                } else if (thumbnailPos == position + 1) {
                    thumbnailPos--;
                }
            }
        }
    }
}

File: app/src/main/java/com/minerdev/greformanager/ImageListAdapter.kt
package com.minerdev.greformanager

import android.content.Context
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import java.io.File
import java.util.*
import kotlin.collections.ArrayList

class ImageListAdapter : RecyclerView.Adapter<ImageListAdapter.ViewHolder>() {
    var imageUris = ArrayList<Uri>()
    var images = ArrayList<Image>()
    private var listener = OnItemClickListener()
    var thumbnail = 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val itemView = inflater.inflate(R.layout.item_image, parent, false)
        return ViewHolder(itemView, listener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(imageUris[position], thumbnail)
    }

    override fun getItemCount(): Int {
        return imageUris.size
    }

    fun addItem(context: Context?, item: Uri) {
        context?.let {
            images.add(addImage(it, item))
            imageUris.add(item)
        }
    }

    fun removeItem(position: Int) {
        images.removeAt(position)
        imageUris.removeAt(position)
    }

    private fun addImage(context: Context, uri: Uri): Image {
        val image = Image()
        val file = File(AppHelper.instance.getPathFromUri(context, uri))
        val fileName: String = file.name
        val fileExtension = fileName.substring(fileName.lastIndexOf(".") + 1)
        image.title = System.currentTimeMillis().toString() + "." + fileExtension
        image.position = imageUris.size.toByte()

        if (image.position == thumbnail.toByte()) {
            image.thumbnail = 1

        } else {
            image.thumbnail = 0
        }

        return image
    }

    class ViewHolder(itemView: View, clickListener: OnItemClickListener?) : RecyclerView.ViewHolder(itemView) {
        private val imageView: ImageView = itemView.findViewById(R.id.imageView)
        private val textView: TextView = itemView.findViewById(R.id.tv_thumbnail)

        fun bind(uri: Uri?, thumbnailPos: Int) {
            Glide.with(itemView).load(uri).into(imageView)

            if (bindingAdapterPosition == thumbnailPos) {
                textView.visibility = View.VISIBLE

            } else {
                textView.visibility = View.GONE
            }
        }

        init {
            val imageBtnDelete = itemView.findViewById<ImageButton>(R.id.image_btn_delete)
            imageBtnDelete.setOnClickListener {
                clickListener?.onDeleteButtonClick(this@ViewHolder, itemView, bindingAdapterPosition)
            }

            val imageBtnThumbnail = itemView.findViewById<ImageButton>(R.id.image_btn_thumbnail)
            imageBtnThumbnail.setOnClickListener {
                clickListener?.onThumbnailButtonClick(this@ViewHolder, itemView, bindingAdapterPosition)
            }

            val imageBtnUp = itemView.findViewById<ImageButton>(R.id.image_btn_up)
            imageBtnUp.setOnClickListener {
                clickListener?.onUpButtonClick(this@ViewHolder, itemView, bindingAdapterPosition)
            }

            val imageBtnDown = itemView.findViewById<ImageButton>(R.id.image_btn_down)
            imageBtnDown.setOnClickListener {
                clickListener?.onDownButtonClick(this@ViewHolder, itemView, bindingAdapterPosition)
            }
        }
    }

    inner class OnItemClickListener {
        fun onDeleteButtonClick(viewHolder: ViewHolder?, view: View?, position: Int) {
            if (thumbnail == position) {
                thumbnail = 0

            } else if (thumbnail > position) {
                thumbnail--
            }

            removeItem(position)
            notifyDataSetChanged()
        }

        fun onThumbnailButtonClick(viewHolder: ViewHolder?, view: View?, position: Int) {
            thumbnail = position
            notifyDataSetChanged()
        }

        fun onUpButtonClick(viewHolder: ViewHolder?, view: View?, position: Int) {
            if (position > 0) {
                Collections.swap(imageUris, position - 1, position)
                Collections.swap(images, position - 1, position)
                notifyDataSetChanged()

                if (thumbnail == position) {
                    thumbnail--

                } else if (thumbnail == position - 1) {
                    thumbnail++
                }
            }
        }

        fun onDownButtonClick(viewHolder: ViewHolder?, view: View?, position: Int) {
            if (position < imageUris.size - 1) {
                Collections.swap(imageUris, position, position + 1)
                Collections.swap(images, position, position + 1)
                notifyDataSetChanged()

                if (thumbnail == position) {
                    thumbnail++

                } else if (thumbnail == position + 1) {
                    thumbnail--
                }
            }
        }
    }
}